<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>MW</class>
 <widget class="QMainWindow" name="MW">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>1072</width>
    <height>759</height>
   </rect>
  </property>
  <property name="windowTitle">
   <string>SimpleMotion Real-time control example</string>
  </property>
  <widget class="QWidget" name="centralWidget">
   <layout class="QHBoxLayout" name="horizontalLayout_2">
    <item>
     <layout class="QVBoxLayout" name="verticalLayout_4">
      <item>
       <widget class="QLabel" name="label_7">
        <property name="text">
         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:large; font-weight:600;&quot;&gt;About this app&lt;/span&gt;&lt;/p&gt;&lt;p&gt;This program implements simple host side position controller over SimpleMotion bus . The program reads actual encoder position from the drive, compares it to user position setpoint, and controls drive's velocity setpoint based on the difference of the encoder vs user setpoint.&lt;/p&gt;&lt;p&gt;The app utilizes smFastUpdateCycle API call to lowest latency and enables highest possible control update rate over SimpleMotion. However, as this program is run on a non-real time OS, it might exhibit some inperfections in response due to control task time jitter. For best results, similar app could be make on a hard real-time system (i.e. RTOS or embedded system).&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:large; font-weight:600;&quot;&gt;Prerequisities&lt;/span&gt;&lt;/p&gt;&lt;p&gt;The app assumes that drive hsa been properly configured and drive is enabled and operating in velocity control mode.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
        </property>
        <property name="wordWrap">
         <bool>true</bool>
        </property>
       </widget>
      </item>
      <item>
       <widget class="QGroupBox" name="groupBox">
        <property name="title">
         <string>Connection</string>
        </property>
        <layout class="QGridLayout" name="gridLayout_3">
         <item row="0" column="0">
          <widget class="QLabel" name="label_2">
           <property name="text">
            <string>SimpleMotion V2 adapter port name (such as COM5, or FTDI0 etc)</string>
           </property>
          </widget>
         </item>
         <item row="0" column="1">
          <widget class="QLineEdit" name="busName">
           <property name="maxLength">
            <number>32</number>
           </property>
          </widget>
         </item>
         <item row="1" column="0">
          <widget class="QLabel" name="label_3">
           <property name="text">
            <string>Target drive address</string>
           </property>
          </widget>
         </item>
         <item row="1" column="1">
          <widget class="QSpinBox" name="deviceAddress">
           <property name="minimum">
            <number>1</number>
           </property>
           <property name="maximum">
            <number>255</number>
           </property>
           <property name="value">
            <number>1</number>
           </property>
          </widget>
         </item>
         <item row="3" column="0">
          <widget class="QPushButton" name="connect">
           <property name="text">
            <string>Connect &amp;&amp; Start control</string>
           </property>
          </widget>
         </item>
         <item row="3" column="1">
          <widget class="QPushButton" name="disconnect">
           <property name="text">
            <string>Stop &amp;&amp; Disconnect</string>
           </property>
          </widget>
         </item>
         <item row="2" column="1">
          <widget class="QCheckBox" name="useHighBaudRate">
           <property name="text">
            <string>Use high baud rate</string>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <widget class="QGroupBox" name="groupBox_2">
        <property name="title">
         <string>Control settings</string>
        </property>
        <layout class="QGridLayout" name="gridLayout">
         <item row="0" column="0">
          <widget class="QLabel" name="label">
           <property name="text">
            <string>Control update rate</string>
           </property>
          </widget>
         </item>
         <item row="0" column="1">
          <widget class="QSpinBox" name="updateRate">
           <property name="suffix">
            <string> Hz</string>
           </property>
           <property name="minimum">
            <number>20</number>
           </property>
           <property name="maximum">
            <number>500</number>
           </property>
           <property name="singleStep">
            <number>10</number>
           </property>
           <property name="value">
            <number>50</number>
           </property>
          </widget>
         </item>
         <item row="1" column="0">
          <widget class="QLabel" name="label_4">
           <property name="text">
            <string>Position gain</string>
           </property>
          </widget>
         </item>
         <item row="3" column="1">
          <widget class="QSpinBox" name="trackingErrorTolerance">
           <property name="suffix">
            <string> encoder counts</string>
           </property>
           <property name="minimum">
            <number>200</number>
           </property>
           <property name="maximum">
            <number>1000000</number>
           </property>
           <property name="singleStep">
            <number>500</number>
           </property>
           <property name="value">
            <number>12000</number>
           </property>
          </widget>
         </item>
         <item row="1" column="1">
          <widget class="QDoubleSpinBox" name="positionGain">
           <property name="maximum">
            <double>9999.989999999999782</double>
           </property>
           <property name="value">
            <double>10.000000000000000</double>
           </property>
          </widget>
         </item>
         <item row="3" column="0">
          <widget class="QLabel" name="label_5">
           <property name="text">
            <string>Position tracking error tolerance</string>
           </property>
          </widget>
         </item>
         <item row="2" column="0">
          <widget class="QLabel" name="label_8">
           <property name="text">
            <string>Maximum velocity setpoint (%)</string>
           </property>
          </widget>
         </item>
         <item row="2" column="1">
          <widget class="QSlider" name="maxSpeed">
           <property name="maximum">
            <number>100</number>
           </property>
           <property name="sliderPosition">
            <number>50</number>
           </property>
           <property name="orientation">
            <enum>Qt::Horizontal</enum>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <widget class="QGroupBox" name="groupBox_4">
        <property name="title">
         <string>Run-time controls</string>
        </property>
        <layout class="QGridLayout" name="gridLayout_2">
         <item row="3" column="0">
          <widget class="QPushButton" name="resetDriveFaults">
           <property name="text">
            <string>Reset drive faults</string>
           </property>
          </widget>
         </item>
         <item row="0" column="0">
          <widget class="QGroupBox" name="groupBox_5">
           <property name="title">
            <string>Adjust position setpoint</string>
           </property>
           <layout class="QVBoxLayout" name="verticalLayout_3">
            <item>
             <layout class="QHBoxLayout" name="horizontalLayout">
              <item>
               <widget class="QPushButton" name="decrement10000">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>-10000</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="decrement1000">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>-1000</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="decrement100">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>-100</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="decrement10">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>-10</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="increment10">
                <property name="sizePolicy">
                 <sizepolicy hsizetype="Minimum" vsizetype="Fixed">
                  <horstretch>0</horstretch>
                  <verstretch>0</verstretch>
                 </sizepolicy>
                </property>
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>+10</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="increment100">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>+100</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="increment1000">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>+1000</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPushButton" name="increment10000">
                <property name="maximumSize">
                 <size>
                  <width>60</width>
                  <height>16777215</height>
                 </size>
                </property>
                <property name="text">
                 <string>+10000</string>
                </property>
               </widget>
              </item>
             </layout>
            </item>
           </layout>
          </widget>
         </item>
         <item row="2" column="0">
          <widget class="QPushButton" name="resetLocalTrackingError">
           <property name="text">
            <string>Reset local position tracking error</string>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <spacer name="verticalSpacer">
        <property name="orientation">
         <enum>Qt::Vertical</enum>
        </property>
        <property name="sizeHint" stdset="0">
         <size>
          <width>20</width>
          <height>40</height>
         </size>
        </property>
       </spacer>
      </item>
     </layout>
    </item>
    <item>
     <layout class="QVBoxLayout" name="verticalLayout">
      <item>
       <widget class="QGroupBox" name="groupBox_6">
        <property name="sizePolicy">
         <sizepolicy hsizetype="Preferred" vsizetype="Maximum">
          <horstretch>0</horstretch>
          <verstretch>0</verstretch>
         </sizepolicy>
        </property>
        <property name="title">
         <string>Status</string>
        </property>
        <layout class="QFormLayout" name="formLayout">
         <item row="1" column="0">
          <widget class="QLabel" name="label_6">
           <property name="text">
            <string>Position setpoint</string>
           </property>
          </widget>
         </item>
         <item row="2" column="0">
          <widget class="QLabel" name="label_9">
           <property name="text">
            <string>Measured position</string>
           </property>
          </widget>
         </item>
         <item row="3" column="0">
          <widget class="QLabel" name="label_11">
           <property name="text">
            <string>Velocity setpoint</string>
           </property>
          </widget>
         </item>
         <item row="1" column="1">
          <widget class="QLCDNumber" name="posSetpoint">
           <property name="digitCount">
            <number>10</number>
           </property>
           <property name="segmentStyle">
            <enum>QLCDNumber::Flat</enum>
           </property>
          </widget>
         </item>
         <item row="2" column="1">
          <widget class="QLCDNumber" name="measuredPosition">
           <property name="digitCount">
            <number>10</number>
           </property>
           <property name="segmentStyle">
            <enum>QLCDNumber::Flat</enum>
           </property>
          </widget>
         </item>
         <item row="3" column="1">
          <widget class="QLCDNumber" name="velocitySetpoint">
           <property name="digitCount">
            <number>10</number>
           </property>
           <property name="segmentStyle">
            <enum>QLCDNumber::Flat</enum>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <widget class="QGroupBox" name="groupBox_3">
        <property name="title">
         <string>Log</string>
        </property>
        <layout class="QVBoxLayout" name="verticalLayout_2">
         <item>
          <widget class="QTextEdit" name="log"/>
         </item>
        </layout>
       </widget>
      </item>
     </layout>
    </item>
   </layout>
  </widget>
  <widget class="QStatusBar" name="statusBar"/>
 </widget>
 <layoutdefault spacing="6" margin="11"/>
 <tabstops>
  <tabstop>busName</tabstop>
  <tabstop>deviceAddress</tabstop>
  <tabstop>useHighBaudRate</tabstop>
  <tabstop>connect</tabstop>
  <tabstop>disconnect</tabstop>
  <tabstop>updateRate</tabstop>
  <tabstop>positionGain</tabstop>
  <tabstop>maxSpeed</tabstop>
  <tabstop>trackingErrorTolerance</tabstop>
  <tabstop>decrement10000</tabstop>
  <tabstop>decrement1000</tabstop>
  <tabstop>decrement100</tabstop>
  <tabstop>decrement10</tabstop>
  <tabstop>increment10</tabstop>
  <tabstop>increment100</tabstop>
  <tabstop>increment1000</tabstop>
  <tabstop>increment10000</tabstop>
  <tabstop>resetLocalTrackingError</tabstop>
  <tabstop>resetDriveFaults</tabstop>
  <tabstop>log</tabstop>
 </tabstops>
 <resources/>
 <connections/>
</ui>
